src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Kafka/KafkaProduceSyncDeliveryHandlerIntegration.cs

```diff
@@ -0,0 +1,179 @@
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Tagging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Kafka
+{
+    /// <summary>
+    /// Confluent.Kafka Producer.TypedDeliveryHandlerShim_Action.HandleDeliveryReport calltarget instrumentation
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = "Confluent.Kafka",
+        TypeName = "Confluent.Kafka.Producer`2+TypedDeliveryHandlerShim_Action",
+        MethodName = ".ctor",
+        ReturnTypeName = ClrNames.Void,
+        ParameterTypeNames = new[] { ClrNames.String, "!0", "!1", KafkaConstants.ActionOfDeliveryReportTypeName },
+        MinimumVersion = "1.4.0",
+        MaximumVersion = "1.*.*",
+        IntegrationName = KafkaConstants.IntegrationName)]
+    public class KafkaProduceSyncDeliveryHandlerIntegration
+    {
+        private static readonly IDatadogLogger Logger = DatadogLogging.GetLoggerFor<KafkaProduceSyncDeliveryHandlerIntegration>();
+
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <typeparam name="TKey">Type of the message key</typeparam>
+        /// <typeparam name="TValue">Type of the message value</typeparam>
+        /// <typeparam name="TActionOfDeliveryReport">Type of the delivery report</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="topic">The topic to which the message was sent</param>
+        /// <param name="key">The message key value</param>
+        /// <param name="value">The message value</param>
+        /// <param name="handler">The delivery handler instance</param>
+        /// <returns>Calltarget state value</returns>
+        public static CallTargetState OnMethodBegin<TTarget, TKey, TValue, TActionOfDeliveryReport>(TTarget instance, string topic, TKey key, TValue value, TActionOfDeliveryReport handler)
+        {
+            if (handler is null)
+            {
+                // Handled in KafkaProduceSyncIntegration
+                return CallTargetState.GetDefault();
+            }
+
+            try
+            {
+                var span = Tracer.Instance?.ActiveScope?.Span;
+                if (span is null)
+                {
+                    Logger.Warning("Unexpected null span for Kafka Producer with delivery handler");
+                    return CallTargetState.GetDefault();
+                }
+
+                var newAction = CachedWrapperDelegate<TActionOfDeliveryReport>.CreateWrapper(handler, span);
+
+                Action<ITypedDeliveryHandlerShimAction> updateHandlerAction = inst => inst.Handler = newAction;
+
+                // store the call to update the handler variable as state
+                // so we update it at the _end_ of the constructor
+                return new CallTargetState(scope: null, state: updateHandlerAction);
```

Comment:
```suggestion
                return new CallTargetState(scope: Tracer.Instance?.ActiveScope, state: updateHandlerAction);
```

Reply:
Is this necessary? You use the previous approach in the BasicGetIntegration! Don't mind making the change, but I assume they're equivalent as we don't use the scope in `MethodEnd`? https://github.com/DataDog/dd-trace-dotnet/blob/a5878ce09f443eea6165a839f09afc674ea0b357/src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/RabbitMQ/BasicGetIntegration.cs#L38

Reply:
Sorry, I should have explained my suggestion. I think you need this change because in the `OnMethodEnd` below, you try to access `state.Scope?.Span.Finish()` but currently `state.Scope` will always be null. So in this case I think it's needed

Reply:
Oh! haha, that's because I moved the try catch to avoid the closure as in your other suggestion. I think actually this flags that by removing the closure I introduced a bug: we can't rely on just fetching the ambient scope here, because we need to ensure we close the span passed to the handler which is _not_ active.

Instead, I think I should remove the `try-catch` from `OnMethodEnd`, and instead move it into the delegate itself (kind of). I think that should work... 🤔